### ISSUE TITLE
[BUG] Dask-cudf merge issues

### ISSUE THREAD

**Issue by ayushdg** [rapidsai/cudf#2694]
**Describe the bug**
There are some correctness issues with dask_cudf merge for specific cases

**Steps/Code to reproduce bug**
### Setup
```
import dask_cudf
import cudf
import numpy as np

df1 = cudf.DataFrame()
df1['a'] = [1,2,3,4,5,6]*2
df1['b'] = np.random.randint(0,12,12)

df2 = cudf.DataFrame()
df2['a'] = [7,2,3,8,5,9]*2
df2['c'] = np.random.randint(0,12,12)

ddf1 = dask_cudf.from_cudf(df1,1)
ddf2 = dask_cudf.from_cudf(df2,1)

ddf1 = ddf1.groupby('a').b.min().to_frame()
ddf2 = ddf2.groupby('a').c.min().to_frame()
```
```
res = ddf1.merge(ddf2,how="left",on=["nonCol"])
res = ddf1.merge(ddf2,how="left",on=["b"])
res = ddf1.merge(ddf2,how="left",on=["c"])
res = ddf1.merge(ddf2,how="left",on=["a"])
```
Ideally this should not work as `nonCol` does not exist in either dataframes, `b` and `c` are specific to only one of the data frames. Note: `a` is the index name in both dfs.
Output:
```
	b	c
a		
2	11	1
3	0	4
5	1	4
1	1	null
4	6	null
6	0	null
```

Seems to be joining on `a` the index.

### Setup (Both df's have same column names)
```
for i in range(5):
    df1 = cudf.DataFrame()
    df1['a'] = [1,2,3,4,5,6]*2
    df1['b'] = np.random.randint(0,12,12)

    df2 = cudf.DataFrame()
    df2['a'] = [7,2,3,8,5,9]*2
    df2['b'] = np.random.randint(0,12,12)

    ddf1 = dask_cudf.from_cudf(df1,1)
    ddf2 = dask_cudf.from_cudf(df2,1)

    ddf1 = ddf1.groupby('a').b.min().to_frame()
    ddf2 = ddf2.groupby('a').b.min().to_frame()

    res = ddf1.merge(ddf2,how="left",on="NonCol")
    res = ddf1.merge(ddf2,how="left",on="a")
    print(res.head(10))
```
Merging on a Non Existent column or index name. Output is different each time based on the input. (Probably not non-deterministic but has to do with the pattern of random numbers that appear/repeat in both dataframes.
```
RESULT
   b
0  2
1  2
2  1
3  0
4  3
5  5
6  1
RESULT
   b
0  8
1  1
2  3
3  5
4  3
5  0
6  8
7  1
RESULT
    b
0   3
1   6
2   4
3   7
4  10
5   1
RESULT
   b
0  1
1  5
2  0
3  0
4  6
5  0
6  1
7  5
8  5
RESULT
   b
0  6
1  6
2  0
3  1
4  3
5  3
```

**Expected behavior**
Host based Dask Dataframe merge operations for the same code above should yield correct output

**Environment overview (please complete the following information)**
 - Environment location: Docker
 - Method of cuDF install: Docker
   - If method of install is [Docker], provide `docker pull` & `docker run` commands used
      0.10 Nightly dev container

**Comment by rjzamora** [comment]
@ayushdg Thanks for raising this.  This problem should be resolved when we remove the dask_cudf `merge` implementation for the one in main-line dask.  I ran some quick tests with experimental branches of cudf/dask_cudf and dask, and cudf-enabled `dask.dataframe.merge` raises errors for the cases that *shouldn't* work.

We should be ready to use `dask.dataframe.merge` in dask_cudf in the near future - The only delay at the moment is that both #2677 and #2655 are needed to avoid device-host memory copying, and I am debugging/preparing dask for these cudf changes in [dask#5322](https://github.com/dask/dask/pull/5322).

**Comment by rjzamora** [comment]
This should be closed by #2878 now

**Comment by ayushdg** [comment]
Awesome! Thanks @rjzamora. Closing this issue
